### PR TITLE
ZO Documentation Test

### PR DESCRIPTION
**Pull request by MarcusHolly**
## Fixes/Resolves:

1st point of Issue #595 

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by MarcusHolly** [comment]
This test is very simple for now, but I'm not sure what other things should be tested besides the two lists matching. That is to say, if the two lists match, then there shouldn't be any other issues pertaining to the documentation.

**Comment by adam-a-a** [comment]
Looking at the test failures very briefly, it seems in a couple of the checks, the rst file list is turning up empty, which leads me to believe that the way you navigate the docs directory may not work as intended. Perhaps you could try adding one/two extra checks that assert the length of each list -- these would just be additional checks to provide more clarity to the developer if ever there were an issue like this in the future, supplementing the list comparison assertion that you already have. However, this doesn't solve your problem with the test failure...

**Comment by adam-a-a** [comment]
Actually- it looks like the "user mode" checks are the ones failing. I am unsure why. @lbianchi-lbl - any thoughts? Is the docs directory path nonexistent when installed as a user? @MarcusHolly you could also add checks to assert that the file paths exist

**Comment by adam-a-a** [comment]
Small suggestion- I'd put the length of list checks before the check to compare lists.

File: watertap/unit_models/zero_order/tests/test_zo_documentation.py
Comment (bknueven) [inline watertap/unit_models/zero_order/tests/test_zo_documentation.py]:
```suggestion
```
I would get rid of this assertion -- the one below will help you debug a missing file, but this one won't be as helpful.

**Comment by bknueven** [comment]
On the specifics of the test failures -- I am surprised the test `test_rst_path` passed on the user-mode builds.

I looked into it, and after installing latest `watertap` release via `pip` in a fresh environment, I *do* see the `docs` directory!
```bash
(watertap_test) bknueven-35255s:zero_order_unit_models bknueven$ pwd
/Users/bknueven/opt/anaconda3/envs/watertap_test/lib/python3.9/site-packages/docs/technical_reference/unit_models/zero_order_unit_models
(watertap_test) bknueven-35255s:zero_order_unit_models bknueven$ ls
__pycache__			automate_rst_file_wt3.py	dummy_sidor_data.yaml
```

I also see our `tutorials` directory and the `bin` directory. Since I think these should be excluded (and end up at the top-level site-packages directory, not inside `watertap`), I think we should address that issue first *before* attempting to fix this PR for user mode tests.

**Comment by MarcusHolly** [comment]
> On the specifics of the test failures -- I am surprised the test `test_rst_path` passed on the user-mode builds.
> 
> I looked into it, and after installing latest `watertap` release via `pip` in a fresh environment, I _do_ see the `docs` directory!
> 
> ```shell
> (watertap_test) bknueven-35255s:zero_order_unit_models bknueven$ pwd
> /Users/bknueven/opt/anaconda3/envs/watertap_test/lib/python3.9/site-packages/docs/technical_reference/unit_models/zero_order_unit_models
> (watertap_test) bknueven-35255s:zero_order_unit_models bknueven$ ls
> __pycache__			automate_rst_file_wt3.py	dummy_sidor_data.yaml
> ```
> 
> I also see our `tutorials` directory and the `bin` directory. Since I think these should be excluded (and end up at the top-level site-packages directory, not inside `watertap`), I think we should address that issue first _before_ attempting to fix this PR for user mode tests.

Alright, thanks for looking at this @bknueven , and your explanation makes sense. I agree that your issue should be addressed first before coming back to fix this PR. So it seems like, after your issue is addressed, the user-mode tests will still fail. Is it better to just bypass these tests or is there a way to/ is it reasonable to install the docs directory directly within the test file (only if it doesnt already exist - which would be the case in user-mode) such that all the tests pass?

As for why the test_rst_path is passing, isn't it because the path itself exists (docs/technical_reference/unit_models/zero_order_unit_models), but the directory only contains the 3 files in your screenshot, none of which are rst files? I would guess that after your issue is addressed, this test would fail because the docs directory will be excluded from the build.

File: setup.py
Comment (lbianchi-lbl) [inline setup.py:92]: Unless there's a strong reason to keep this, we should consider reverting this change before merging.

File: watertap/unit_models/zero_order/tests/test_zo_documentation.py
Comment (lbianchi-lbl) [inline watertap/unit_models/zero_order/tests/test_zo_documentation.py]: I think the reason why the tests are failing is that `doc_path` exists (so the test doesn't get skipped), but `rst_path` doesn't (so the test fails). Therefore, it might work with this change:

```suggestion
@pytest.mark.skipif(not os.path.exists(rst_path), reason="Directory with .rst source files for ZO models docs not found")
```

**Review by lbianchi-lbl — CHANGES_REQUESTED** [review]
I've left a suggestion on a possible way to address the test failures in the non-editable ("user mode") CI environment.

Apart from that, I think the current implementation is OK, but we might want to revisit it to make it more robust sooner rather than later. More in general, this test has two aspects that make it tricky:

- Matching between the Sphinx docs and the Python code (including e.g. finding the location of the `docs` directory, that might or might not be there depending on the installation mode)
- Finding all ZO models, which in turns requires listing all Python modules within a given package

Both of these would be useful to have for other tests beyond this one, so it might make sense to merge this (as soon as it works reliably) and then refactor it together with the other related tests in a separate PR.

File: setup.py
Comment (lbianchi-lbl) [reply setup.py:92]: To clarify:

- I think the point being made in #879 is valid, and we should modify the `setup.py` settings so that files and directories other than `watertap` don't end up in the installation
- However, since this change would impact other things beyond the original scope of this PR (and I'm planning to rework our distribution setup soon), it might make sense to deal with it separately
- That said, ultimately I don't have strong preferences and I'm reasonably confident that our tests would catch any possible regression introduced by this change, so I'd be OK if we wanted to merge it as part of this PR

File: watertap/unit_models/zero_order/tests/test_zo_documentation.py
Comment (lbianchi-lbl) [reply watertap/unit_models/zero_order/tests/test_zo_documentation.py]: On second thought, reading https://github.com/watertap-org/watertap/pull/877#issuecomment-1371266240 and https://github.com/watertap-org/watertap/pull/877#issuecomment-1371293292, I don't think this would work: the directory would exist, but without any `.rst` file.

Testing for the presence of at least one `.rst` file should work:

```suggestion
doc_path = Path(doc_path).resolve()

...

@pytest.mark.skipif(not os.path.exists(doc_path / "index.rst"), reason="No RST files found")
```

However, this starts to look a bit too brittle for my tastes...

File: setup.py
Comment (MarcusHolly) [reply setup.py:92]: Although it may look a bit brittle, all of the tests are passing now. As for this change, I'll let @bknueven comment on if there's a strong reason to keep it.

Comment (lbianchi-lbl) [reply setup.py:92]: Yeah, I agree, I don't think there's any harm in merging this as-is (including the change in `setup.py`).

**Review by lbianchi-lbl — APPROVED** [review]
As far as I'm concerned, we can merge this as it is and then refactor the shared functionality in a separate PR.

**Review by bknueven — APPROVED** [review]

**Pull request merged**
